### DISCUSSION TITLE
Collision avoidance Python API - too awkward?

### DISCUSSION THREAD

**Discussion by jpreiss** [USC-ACTLab/crazyswarm#481] [Ideas]
To enable collision avoidance programmatically in the script (as opposed to in the launch file), the user must pass a list of the other `Crazyflie` objects to avoid:

https://github.com/USC-ACTLab/crazyswarm/blob/73cce6bb2b33c4fd5415713d25efd35d3f208b08/ros_ws/src/crazyswarm/scripts/collisionAvoidance.py#L71-L73

The simulator does not parse the launch file, so in practice this is the *only* way to enable collision avoidance, unless you're OK running your multi-robot script on hardware without ever testing in simulation.

The API is like this because `Crazyflie` objects in the simulator don't know about each other.

If we had instead given each `Crazyflie` a pointer back to the `CrazyflieServer`, we wouldn't have needed the `others` argument. This would be more user-friendly. But now removing the `others` argument would break the API because there is another positional argument after it.

On the other hand, the argument could be an opportunity to catch bugs: we could add a check that all the `others` are on the same radio channel as `self`. Technically it is not wrong to have two separate collision avoidance groups on different channels (as long as the groups are guaranteed not to collide with each other by other means), so the `others` argument is the most general way to verify that the radio channels match the user's intent.

Opinions?

**Comment by whoenig** [comment]
Another option would be to have an API on the `swarm` object, not per individual Crazyflie. Then, we could warn in the physical drone case if more than one channel is used. We could also deprecate the existing per-CF function.

**Reply by jpreiss** [reply]
I think there are use cases for enabling it at a per-CF level - consider something like leader/follower flocking.

I like the idea of using warnings to catch possible user mistakes without adding user code complexity. We can still do that in the per-CF function if we give each `Crazyflie` a pointer to the `CrazyflieServer`.
